src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,225 @@
+                if( getEditDistance(uniqueObservedUMIs.get(i), uniqueObservedUMIs.get(j)) <= editDistanceToJoin) {
+                    adjacencyList.get(i).add(j);
+                }
+            }
+        }
+
+        // Join Groups
+        int nGroups = 0;
+        for(int i = 0;i < adjacencyList.size();i++) {
+            // Have I visited this yet?
+            if(groups.get(i) == 0) {
+                // No, I haven't yet seen this
+                nGroups++; // We've now seen a new group
+
+                // Depth first search on adjacencyList, setting all the values to a group
+                mergeGroups(i, nGroups);
+            }
+        }
+
+        // Construct DuplicateSetList
+        List<DuplicateSet> duplicateSetList= new ArrayList<>();
+        for(int i = 0;i < nGroups;i++) {
+            DuplicateSet e = new DuplicateSet();
+            duplicateSetList.add(e);
+        }
+
+        // Assign each record to a duplicate set
+        for(int i = 0;i < records.size();i++) {
+            String umi = (String) records.get(i).getAttribute("RX");
+
+            // Figure out which group this belongs to
+            int recordGroup = groups.get(uniqueObservedUMIs.indexOf((String) umi));
+            duplicateSetList.get(recordGroup-1).add(records.get(i));
+
+        }
+
+        // Optionally add the inferred Umi
+        if(addInferredUmi) {
+            for(int i = 0;i < nGroups;i++) {
+                // For each duplicate set identify the most common UMI
+                List<SAMRecord> recordsFromDuplicateSet = duplicateSetList.get(i).getRecords();
+
+                // Count the number of times each UMI appears
+                Map<String, Long> umiCounts = new HashMap<>();
+                for(SAMRecord record : recordsFromDuplicateSet) {
+                    String currentUmi = (String) record.getAttribute("RX");
+                    Long currentCount = umiCounts.get(currentUmi);
+                    if(currentCount == null) {
+                        umiCounts.put(currentUmi, new Long(1));
+                    } else {
+                        umiCounts.put(currentUmi, currentCount + 1);
+                    }
+                }
+
+                // Find most common UMI
+                Map.Entry<String, Long> maxEntry = null;
+                for(Map.Entry<String, Long> entry : umiCounts.entrySet()) {
+                    if(maxEntry == null || entry.getValue().compareTo(maxEntry.getValue()) > 0) {
+                        maxEntry = entry;
+                    }
+                }
+
+                // Assign the inffered UMI to all reads in the current group
+                for(SAMRecord record : recordsFromDuplicateSet) {
+                    record.setAttribute("RI", maxEntry.getKey());
+                }
+            }
+        }
+
+        nextSetsIterator = duplicateSetList.iterator();
+    }
+
+    private void mergeGroups(final int index, final int group) {
```

Comment:
(easier, I think, if groups would be an argument to this function rather than a global variable)